### PR TITLE
add mathematical companion

### PR DESCRIPTION
**Pull request by apoelstra**
Adds a first draft of the "mathematical companion" document, which is updated with new photos of the volvelles, correct descriptions of them, and new sections on quickchecks and error correction.

### CODE REVIEW

File: mathematical-companion/main.tex [hunk 1]
```diff
@@ -0,0 +1,1300 @@
...
+all codewords must must have modulo $G$. All $m$ values are equivalent, in
+the sense that it is easy to convert from one $m$ value to another (simply
+add the difference to the checksum characters of every codeword).
+
+But $m=0$ has the particularly bad property that any codeword can be extended
+by addition of an arbitrary number of 0s to get another valid codeword. (This
+highlights the fact that BCH codes are designed to handle only \emph{substitution}
+or \emph{erasure} errors, not insertions or deletions.) Other small values of
+$m$ have similar issues; bech32 was originally defined to have $m=1$ but later
+needed to be modified to bech32m for this reason. bech32m uses a large random
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:798]: not actually random.

File: mathematical-companion/main.tex [hunk 2]
```diff
@@ -0,0 +1,1300 @@
...
+as a time: multiply by $x^2$, add two new characters, reduce, repeat. In detail:
+
+\paragraph{Checksum Verification.}
+
+\begin{enumerate}
+\item Take the prefix \vc{ms1} and encode it as a bech32 human readable part (HRP),
+which works out to \vc{RRQDN} (the process is to take the high 3 bits of each
+8-bit ASCII character, followed by 0, followed by the low 5 bits of each character).
+Prefix a 1, or \vc{P} in bech32. The resulting initial polynomial is \vc{PRRQDN}, or
+\[ x^5 + \binrep{3}x^4 + \binrep{3}x^3 + 0x^2 + \binrep{13}x + \binrep{19} \]
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:896]: Use Bech32 characters here.

File: mathematical-companion/main.tex [hunk 3]
```diff
@@ -0,0 +1,1300 @@
...
+
+\item During recovery, there are up to 31 outstanding non-\vc{S} shares and
+the user has an arbitrary subset of $k$ of them. There are $\binom{31}k$
+possibilities, which for $k=2$ is a reasonable number --- 465 --- but for
+$k\geq3$ quickly grows out of hand, to 4495, then 31465, and beyond.
+
+For $k = 2$ we provide a table, both in tabular form and in the form of a volvelle.
+This is what the Recovery Wheel ultimately is, and at position $p$, window $w$ it
+computes
+\[ \ell_i(\vc{S}) = \frac{w + \vc{S}}{w + p} \]
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:999]: `\ell_i(w)`

File: mathematical-companion/main.tex [hunk 4]
```diff
@@ -0,0 +1,1300 @@
...
+Here position $p$ is the index of share $y_i$ whose multiplier is being computed,
+and $w$ is the index of the other share. To avoid getting $p$ and $w$ confused,
+we have written ``Share to Translate'' on the handle of the disc.
+
+For $k \geq 3$, for target share $p$ (index of $y_i$) with other shares
+$\{w_i\}_{i=1}^{k-1}$, we need to compute
+\[ \ell_i(\vc{S}) = \prod_{i=1}^{k-1} \frac{w_i + \vc{S}}{w_i + p} \]
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:1006]: `\ell_i(w) = \prod_{i=1}^{k-1} \frac{w + \vc{S}}{w + p_i}`

File: mathematical-companion/main.tex [hunk 5]
```diff
@@ -0,0 +1,1300 @@
...
+next, we need to dig a bit deeper into how it was constructed.
+
+To this end we first construct the extension field $\ftttwo$. Just as we
+produced $\fttwo$ by adding a root of the irreducible 5th-degree polynomial
+$x^5 + x^3 + 1$ over $\ftwo$, we can adjoin a root of the irreducible
+2nd-degree polynomial $x^2 + x + 1$ to $\fttwo$ to get a new field $\ftttwo$.
+We will call this new root $\zeta$. By construction, $\zeta$ satisfies
+the equation $\zeta^2 = \zeta + 1$.
+
+We can write any element of $\ftttwo$ as $a\zeta + b$, where $a$ and $b$ are
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:1099]: I think `a + b\zeta` would be more traditional.  I guess I could be wrong.

File: mathematical-companion/main.tex [hunk 6]
```diff
@@ -0,0 +1,1300 @@
...
+order in the set $\{3, 11, 31, 33, 93, 341, 1023\}$.
+
+You may recognize the numbers 93 and 1023 as the length of our primary and
+alternate code (and 1023 as the length of bech32). This is not a coincidence.
+
+Consider the element $\beta = \vc{G}\zeta$, which has order $93$. In fact, our
+generator $G$ has roots which are all powers of $\beta$! We can write it as
+\begin{align*}
+    G(x) = \prod_{i\in\{17, 20, 46, 49, 52, 77, 78, 79, 80, 81, 82, 83, 84\}} (x - \beta^i).
+\end{align*}
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:1115]: A slightly more honest phrasing is that we took the minimal polynomials of \beta^77 … \beta^84 and multiplied them all together.  That does require you to define what a minimal polynomial is thought.

File: mathematical-companion/main.tex [hunk 7]
```diff
@@ -0,0 +1,1300 @@
...
+back over with~\eqref{eq:quickstart}.
+
+You may notice that the factor $(x + \vc{S}) = (x - \beta^{81})$ appears twice in
+this list. For a user doing all the checks, its second appearance is strictly
+redundant. But by including it, we ensure that no matter where in the list the
+user is starting from, they consistently accumulate consecutive roots, so that
+if their data isn't corrupted between individual checks, they gain progressive
+error detection ability.
+
+Another reason to include $(x+\vc{S})$ twice is to make all the quickchecks look
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:1183]: The most important reason is that a quickcheck detects any other types of errors with 99.9% probability (1023/ 1024).  If we only did a linear check that would reduce the detection probability to 31/32.

File: mathematical-companion/main.tex [hunk 8]
```diff
@@ -0,0 +1,1300 @@
...
+corrupted. An error in location $y$ actually means that \emph{column $y$ in the
+Checksum Worksheet was not computed correctly}. This may mean that the top cell,
+which contains share data, was wrong. But it could also be that the user made an
+arithmetic error. We therefore advise users, when doing error correction, to
+completely recompute any erroneous columns of the Checksum Worksheet before they
+consider modifying their data.
+
+\section{Conclusion and Acknowledgements}
+
+We thank the authors of SLIP39 for noticing the remarkable compatibility between
```
Comment (roconnor-blockstream) [inline mathematical-companion/main.tex:1260]: AFAIK, the authors of SLIP39 didn't notice this compatibility. It was Leon.

File: mathematical-companion/main.tex (hunk 6 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:1115]: I think I already talk about the "degree of a field element", which I may not have defined (it is the degree of the element's minimal polynomial). So I'll make sure there's a definition of minimum polynomial then reword this.

File: mathematical-companion/main.tex (hunk 8 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:1260]: Ok, I'll update this text.

File: mathematical-companion/main.tex [hunk 9]
```diff
@@ -0,0 +1,1300 @@
...
+We call such a set a \textbf{basis} for $\fttwo$.
+
+\subsection{Lagrange Interpolation and Shamir's Secret Sharing\label{sec:sss}}
+
+Let $\mathbb{F}$ be a field and $p$ a polynomial of degree $n$ in $\mathbb{F}[x]$. It is a
+standard theorem of algebra that $p$'s value on all points of $\mathbb{F}$ is
+implied by its values on any $n+1$ distinct points.
+
+As discovered by Edward Waring in 1779, and later by Joseph-Louis Lagrange
+in 1795\footnote{citation: Wikipedia}, it is actually possible to compute
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply mathematical-companion/main.tex:154]: I'll try to, I guess by just using whatever Wikipedia cites. But it's an obscure and non-essential fact so I think it's fine to be a bit silly regarding its citation.

File: mathematical-companion/main.tex (hunk 1 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:798]: Do you know offhand how it was chosen? I may just say "random-looking" instead.

File: mathematical-companion/main.tex (hunk 5 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:1099]: Hmm, yeah, that is how we write complex numbers isn't it. I'll go over the doc and reverse the ordering here.

File: mathematical-companion/main.tex (hunk 1 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:798]: Ah, sipa has a gist at https://gist.github.com/sipa/14c248c288c3880a3b191f978a34508e

File: mathematical-companion/main.tex (hunk 3 above)
Comment (apoelstra) [reply mathematical-companion/main.tex:999]: I think it should be `ell_i(S)`. The idea is that `\ell_i(x)` is a polynomial in an indeterminate `x` ... and we are evaluating it at `S` when doing recovery.

**Review by roconnor-blockstream — APPROVED** [review]
LGTM

**Comment by apoelstra** [comment]
Great!

I'll let you hit "merge" on this, unless you want me to.

**Pull request merged**